mlflow/tensorflow/_autolog.py

```diff
@@ -52,11 +59,162 @@ def on_epoch_end(self, epoch, logs=None):
+        elif isinstance(features, (np.ndarray, tensorflow.Tensor)):
+            return _extract_input_example_from_tensor_or_ndarray(features)
+        raise TypeError(f"Unsupported type for features of type: {type(features)}")
+    raise TypeError(f"Unsupported type for dataset batch slice of type: {type(first_batch)}")
+
+
+def extract_data_from_tf_input_fn(input_fn):
+    """
+    Extracts sample data from dict (str -> ndarray),
+    ``tensorflow.Tensor`` or ``tensorflow.data.Dataset`` type.
+
+    :param input_fn: Tensorflow's input function used for train method
+    :return: a slice (of limit ``mlflow.utils.autologging_utils.INPUT_EXAMPLE_SAMPLE_ROWS``)
+             of the input of type `np.ndarray`
+    """
+
+    input_training_data = input_fn()
+    input_features = None
+    if isinstance(input_training_data, tuple):
+        features = input_training_data[0]
+        if isinstance(features, dict):
+            input_features = _extract_sample_numpy_dict(features)
+        elif isinstance(features, (np.ndarray, tensorflow.Tensor)):
+            input_features = _extract_input_example_from_tensor_or_ndarray(features)
+    elif isinstance(input_training_data, tensorflow.data.Dataset):
+        input_features = _extract_input_example_from_batched_tf_dataset(input_training_data)
+    return input_features
+
+
+def get_tf_keras_input_data_slice(input_training_data):
+    """
+    Generates a sample ndarray or dict (str -> ndarray)
+    from the input type 'x' for keras ``fit`` or ``fit_generator``
+
+    :param input_training_data: Keras input function used for ``fit`` or
+                                ``fit_generator`` methods
+    :return: a slice of type ndarray or
+             dict (str -> ndarray) limited to
+             ``mlflow.utils.autologging_utils.INPUT_EXAMPLE_SAMPLE_ROWS``
+    """
+
+    if isinstance(input_training_data, tensorflow.keras.utils.Sequence):
+        input_training_data = input_training_data[:][0]
+
+    if isinstance(input_training_data, (np.ndarray, tensorflow.Tensor)):
+        input_data_slice = _extract_input_example_from_tensor_or_ndarray(input_training_data)
+    elif isinstance(input_training_data, dict):
+        input_data_slice = _extract_sample_numpy_dict(input_training_data)
+    elif isinstance(input_training_data, tensorflow.data.Dataset):
+        input_data_slice = _extract_input_example_from_batched_tf_dataset(input_training_data)
+    else:
+        raise MlflowException(
+            "Cannot log input example or model signature for input with type"
+            f" {type(input_training_data)}. TensorFlow Keras autologging can"
+            " only log input examples and model signatures for the following"
+            " input types: numpy.ndarray, dict[string -> numpy.ndarray],"
+            " tensorflow.keras.utils.Sequence, and"
+            " tensorflow.data.Dataset (TensorFlow >= 2.1.0 required)",
+            INVALID_PARAMETER_VALUE,
+        )
```

Comment:
Can we move this exception into the part of the code in `mlflow.tensorflow.autolog()` where the input example / model signature is resolved? This utility method is just responsible for trying to get the data for the caller. It's up to the caller how to handle the case where the data is `None`, so the caller should throw.